transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,697 @@
+    Any { if_watch: IfWatch },
+}
+
+/// A WebRTC transport with direct p2p communication (without a STUN server).
+#[derive(Clone)]
+pub struct WebRTCTransport {
+    /// A `RTCConfiguration` which holds this peer's certificate(s).
+    config: RTCConfiguration,
+
+    /// The `UDPMux` that manages all ICE connections.
+    udp_mux: Arc<dyn UDPMux + Send + Sync>,
+
+    /// The local address of `udp_mux`.
+    udp_mux_addr: SocketAddr,
+
+    /// The receiver for new `SocketAddr` connecting to this peer.
+    new_addr_rx: Arc<Mutex<mpsc::Receiver<Multiaddr>>>,
+}
+
+impl WebRTCTransport {
+    /// Create a new WebRTC transport.
+    ///
+    /// Creates a UDP socket bound to `listen_addr`.
+    pub async fn new<A: ToSocketAddrs>(
+        certificate: RTCCertificate,
+        listen_addr: A,
+    ) -> Result<Self, TransportError<Error>> {
+        // bind to `listen_addr` and construct a UDP mux.
+        let socket = UdpSocket::bind(listen_addr)
+            .map_err(Error::IoError)
+            .map_err(TransportError::Other)
+            .await?;
+        // sender and receiver for new addresses
+        let (new_addr_tx, new_addr_rx) = mpsc::channel(1);
+        let udp_mux_addr = socket
+            .local_addr()
+            .map_err(Error::IoError)
+            .map_err(TransportError::Other)?;
+        let udp_mux = UDPMuxNewAddr::new(UDPMuxParams::new(socket), new_addr_tx);
+
+        Ok(Self {
+            config: RTCConfiguration {
+                certificates: vec![certificate],
+                ..Default::default()
+            },
+            udp_mux,
+            udp_mux_addr,
+            new_addr_rx: Arc::new(Mutex::new(new_addr_rx)),
+        })
+    }
+
+    /// Returns the SHA-256 fingerprint of the certificate in lowercase hex string as expressed
+    /// utilizing the syntax of 'fingerprint' in <https://tools.ietf.org/html/rfc4572#section-5>.
+    fn cert_fingerprint(&self) -> String {
+        fingerprint_of_first_certificate(&self.config)
+    }
+}
+
+impl Transport for WebRTCTransport {
+    type Output = Connection<'static>;
+    type Error = Error;
+    type Listener = WebRTCListenStream;
+    type ListenerUpgrade = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+    type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+        debug!("listening on {} (ignoring {})", self.udp_mux_addr, addr);
+        Ok(WebRTCListenStream::new(
+            self.udp_mux_addr,
+            self.config.clone(),
+            self.udp_mux.clone(),
+            self.new_addr_rx.clone(),
+        ))
+    }
+
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        Ok(Box::pin(self.do_dial(addr)))
+    }
+
+    fn dial_as_listener(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        // XXX: anything to do here?
+        self.dial(addr)
```

Comment:
This is required in the case of hole punching. (See below.) Eventually we might want a browser to be able to hole punch to a rust-libp2p server node. I am not yet sure how this concept would then translate into the WebRTC world. Will need to give this more thought.

> 5. Simultaneous Connect. The two nodes follow the steps below in parallel for
>    every address obtained from the `Connect` message:
>    - For a TCP address:
>       - Upon receiving the `Sync`, `A` immediately dials the address to `B`.
>       - Upon expiry of the timer, `B` dials the address to `A`.
>       - This will result in a TCP Simultaneous Connect. For the purpose of all
>         protocols run on top of this TCP connection, `A` is assumed to be the
>         client and `B` the server.
>    - For a QUIC address:
>       - Upon receiving the `Sync`, `A` immediately dials the address to `B`.
>       - Upon expiry of the timer, `B` starts to send UDP packets filled with
>         random bytes to `A`'s address. Packets should be sent repeatedly in
>         random intervals between 10 and 200 ms.
>       - This will result in a QUIC connection where `A` is the client and `B` is
>         the server.

https://github.com/libp2p/specs/blob/master/relay/DCUtR.md